### PR TITLE
fix(wallet): hide buy for unsupported tokens

### PR DESCRIPTION
**Pull request by Pavneet-Sing**
Resolves https://github.com/brave/brave-browser/issues/27876

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

[hide-buy-asset.webm](https://user-images.githubusercontent.com/11755381/213190482-b7078691-277f-418c-a300-0cb42c7f7c80.webm)

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java [hunk 1]
```diff
@@ -111,6 +112,10 @@ protected void triggerLayoutInflation() {
         }
         mExecutor = Executors.newSingleThreadExecutor();
         mHandler = new Handler(Looper.getMainLooper());
+        BraveActivity activity = BraveActivity.getBraveActivity();
+        if (activity != null) {
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: Considering the `mWalletModel` field is used in the `observeOnce` method below, it's best to assert (i.e. `assert activity != null`) to be sure the Brave activity will never be `null`. Very unlikely to happen if I've understood correctly, but asserting we make sure we don't access a null object.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AssetUtils.java [hunk 2]
```diff
@@ -51,6 +51,20 @@ public class AssetUtils {
...
+        public static boolean isSameToken(BlockchainToken token1, BlockchainToken token2) {
+            return token1.symbol.equals(token2.symbol)
+                    && token1.contractAddress.equals(token2.contractAddress)
+                    && token1.chainId.equals(token2.chainId);
+        }
+
+        public static boolean isSameToken(
+                BlockchainToken token1, String symbol, String contractAddress, String chainId) {
+            return token1.symbol.equals(symbol) && token1.contractAddress.equals(contractAddress)
+                    && token1.chainId.equals(chainId);
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/AssetUtils.java:64]: Nit:

```suggestion
                BlockchainToken token, String symbol, String contractAddress, String chainId) {
            return token.symbol.equals(symbol) && token.contractAddress.equals(contractAddress)
                    && token.chainId.equals(chainId);
```

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java [hunk 3]
```diff
@@ -37,4 +37,30 @@ public static <T> List<T> filter(List<T> list, Predicate<T> filter) {
         }
         return filteredList;
     }
+
+    public static <T> T find(List<T> list, Predicate<T> predicate) {
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: Nit: `find` -> `findFirst`, so we match the Java style find pattern.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java [hunk 4]
```diff
@@ -37,4 +37,30 @@ public static <T> List<T> filter(List<T> list, Predicate<T> filter) {
...
+            }
+        }
+        return null;
+    }
+
+    public static <T> boolean includes(List<T> list, Predicate<T> predicate) {
+        return find(list, predicate) != null;
+    }
+
+    public static <T> T find(T[] items, Predicate<T> predicate) {
```
Comment (simoarpe) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:54]: Nit: `find` -> `findFirst`, so we match the Java style find pattern.

**Review by simoarpe — APPROVED** [review]
👍 Added just a couple of suggestions. Code looks good to me.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java (hunk 1 above)
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: no, without assert the app will crash anyway so it is not required here.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java (hunk 3 above)
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: we don't have to match, `find` means finding the first matched object and java doesn't have any `find`.

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java (hunk 1 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: So if we are not preventing the app from crashing, what's the point of this check? 🤔

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java (hunk 3 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: In Java 8, they introduced streams, and there is `Stream.findAny()` and `Stream.findFirst()`. It was introduced to bring clarity on the actions performed.

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: Adding a few links here, but just for reference:
* Oracle documentation: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#findFirst--
* Find first API explaination: https://stackify.com/streams-guide-java-8/

Anyway, it's just nitpicking. Feel free to adopt what you feel more comfortable with.

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: The point is we are not mimicing Streams or Java8 APIs and there is no need. We can choose to be specific when it is required and in this case it is not.

File: android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java (hunk 1 above)
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: how assert will help here to prevent a crash?

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: maybe we can just add a check on `mWalletModel == null` below and do nothing for now. That should go away when we move everything out of the activity as we discussed in a DM

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/activities/AssetDetailActivity.java:116]: I can add the null check though as per the logs, that seems to be the issue with the internal `BraveWalletService_Internal.MANAGER` object being null and outer objects are initialised properly. Gonna add the check.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java (hunk 3 above)
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: Alright, a Javadoc, or a short comment surely will help to clarify and prevent any confusion.

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: What confusion?

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: It's always good practice to include Javadoc comments on public methods. Not mandatory, but it's even more useful for util classes that could be used in so many places by different team members.
Now everything is fresh and clear, but in a couple of weeks, or months, a small indication of what a method does won't harm, and will make the code easier to follow.

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/JavaUtils.java:41]: Not always a good practice to add docs, it depends on many things(time, type, implementation etc.) though I have been adding docs when it is required.

**Comment by srirambv** [comment]
Verification passed on Lenovo Tab running 1.49.41 x64 nightly build
- Verified steps from PR and issue description
- Verified `Buy` button is not shown for unsupported tokens 

https://user-images.githubusercontent.com/17010094/213384765-777e1535-ca05-4dc7-b426-87fea88bd6bc.mp4

**Pull request merged**